### ISSUE TITLE
Test on webcam

### ISSUE THREAD

**Issue by DaddyWesker** [JiaweiShiCV/Amend-Representation-Module#14]
Hello! THanks for your code.

I've tried to make a python script to use your model with webcam but it seems i'm using wrong labels. Can you help me with that? Here is the script.

```
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import torch.utils.data as data
from torchvision import transforms
import torch
import argparse
import Networks
from sklearn.metrics import confusion_matrix
from plot_confusion_matrix import plot_confusion_matrix
import cv2
from rmn import RMN

model = Networks.ResNet18_ARM___RAF()
print("Loading pretrained weights...", "epoch59_acc0.9205.pth")
checkpoint = torch.load("epoch59_acc0.9205.pth")
model.load_state_dict(checkpoint["model_state_dict"], strict=False)
model = model.cuda()

video_capture = cv2.VideoCapture(0)
m = RMN()

labels_name = ['surprise', 'fear', 'disgust', 'happy', 'sad', 'angry', "neutral"]
with torch.no_grad():
    bingo_cnt = 0
    model.eval()
    while True:
        _, frame = video_capture.read()
        face = m.detect_faces(frame)
        xmin = face[0]["xmin"]
        ymin = face[0]["ymin"]
        xmax = face[0]["xmax"]
        ymax = face[0]["ymax"]
        bgr_frame = frame
        frame = frame[:, :, ::-1]
        face_image = frame[ymin:ymax, xmin:xmax]

        face_image = cv2.resize(face_image, (196, 196), interpolation=cv2.INTER_CUBIC)
        face_image = np.transpose(face_image, (2, 0, 1))
        t_face_image = torch.FloatTensor(face_image.copy()).cuda()
        t_face_image = torch.unsqueeze(t_face_image, 0)
        outputs, _ = model(t_face_image)
        _, predicts = torch.max(outputs, 1)
        pred = labels_name[predicts.cpu().numpy()[0]]
        cv2.putText(bgr_frame, pred, (30, 50), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (255, 0, 0), 2)
        cv2.imshow("emotion", bgr_frame)
        cv2.waitKey(1)
```
 Thing is that i'm using 
`labels_name = ['surprise', 'fear', 'disgust', 'happy', 'sad', 'angry', "neutral"]` which is probably wrong, since when i'm neutral model says that i'm sad. Or am i using it somehow wrong?

**Comment by JiaweiShiCV** [comment]
The same issue as #9. Since our SA block utilizes global average features in a batch, the recognition depends on the batch size of 64. It is temporarily unable to support the recognition of a single picture unless the SA block is removed during training.

**Comment by DaddyWesker** [comment]
So, if i got you right, if i want to test on webcam i need to collect 64 consequent frames and send them as one batch?

**Comment by JiaweiShiCV** [comment]
Yes

**Comment by DaddyWesker** [comment]
ALright. My code looks like that now
```
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import torch.utils.data as data
from torchvision import transforms
import torch
import argparse
import Networks
from sklearn.metrics import confusion_matrix
from plot_confusion_matrix import plot_confusion_matrix
import cv2
from rmn import RMN

model = Networks.ResNet18_ARM___RAF()
print("Loading pretrained weights...", "epoch59_acc0.9205.pth")
checkpoint = torch.load("epoch59_acc0.9205.pth")
model.load_state_dict(checkpoint["model_state_dict"], strict=False)
model = model.cuda()

video_capture = cv2.VideoCapture(0)
m = RMN()

labels_name = ['surprise', 'fear', 'disgust', 'happy', 'sad', 'angry', "neutral"]
with torch.no_grad():
    bingo_cnt = 0
    model.eval()
    frames = []
    while True:
        _, frame = video_capture.read()
        face = m.detect_faces(frame)
        xmin = face[0]["xmin"]
        ymin = face[0]["ymin"]
        xmax = face[0]["xmax"]
        ymax = face[0]["ymax"]
        bgr_frame = frame
        frame = frame[:, :, ::-1]
        face_image = frame[ymin:ymax, xmin:xmax]

        face_image = cv2.resize(face_image, (196, 196), interpolation=cv2.INTER_CUBIC)
        face_image = np.transpose(face_image, (2, 0, 1))
        t_face_image = torch.FloatTensor(face_image.copy()).cuda()
        t_face_image = torch.unsqueeze(t_face_image, 0)
        frames.append(t_face_image)
        if (len(frames) < 64):
            continue
        outputs, _ = model(torch.cat(frames))
        _, predicts = torch.max(outputs, 1)
        predicts = predicts.cpu().numpy()
        all_labels = [0,0,0,0,0,0,0]
        for i in range(len(predicts)):
            all_labels[predicts[i]] += 1
        pred = labels_name[np.argmax(all_labels)]
        cv2.putText(bgr_frame, pred, (30, 50), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (255, 0, 0), 2)
        cv2.imshow("emotion", bgr_frame)
        cv2.waitKey(1)
        frames.pop(0)
```
but still, model's predict is label "4" (sad), even when i'm smiling for 64 consequent frames. Moreover, now "4" is the only prediction. Maybe i'm doing something wrong?

**Comment by lucaskyle** [comment]
@DaddyWesker 

some problems in ur code.

1.  the input of  Networks.ResNet18_ARM___RAF() should be (64,3,224,224) or (1,3,224,224)    
> face_image = cv2.resize(face_image, (196, 196), interpolation=cv2.INTER_CUBIC)

2. also u need normalize the input face image according to the training phase.
like:

> mean=[0.485, 0.456, 0.406]
> std=[0.229, 0.224, 0.225]
> face_image = ((face_image / 255. - mean) / std).astype(np.float32)

3. put a threshold to control the value of the outputs instead of using torch.max() or np.argmax()

**Comment by DaddyWesker** [comment]
@lucaskyle 

THanks for your tips! Tip 1 and 2 helped me to make it work better, though i don't quite get it what sort of threshold you mean and on which value? I'm getting outputs as a tensor [64,7]. DO you mean applying threshold to each value of this tensor? But i'm still need to find max value for each 64 of 7-dim arrays to get max emotion for each frame. Or not? I'll take a look into train and test codes, maybe i''ve missed something.
